src/main/java/com/es/phoneshop/service/cart/DefaultCartService.java

```diff
@@ -48,9 +51,55 @@ public void addProductToCart(Cart cart, Long productId, int quantity) throws Out
             } else {
                 cart.getItems().add(new CartItem(product, quantity));
             }
+            recalculateCart(cart);
         }
     }
 
+    @Override
+    public void updateProductInCart(Cart cart, Long productId, int quantity) throws OutOfStockException, ProductNotFoundException {
+        Object syncObject = SyncObjectPool.getSyncObject(cart.getId().toString());
+        synchronized (syncObject) {
+            Product product = productService.getProductById(productId);
+            Optional<CartItem> optionalCartItem = findCartItem(cart, product);
+            checkQuantity(product, quantity, 0);
+
+            if (optionalCartItem.isPresent()) {
+                if (optionalCartItem.get().getQuantity() == quantity) {
+                    throw new SameArgumentException(optionalCartItem.get());
```

Comment:
we can just ignore those products which should not be updated, exceptions are necessary for the exceptional cases
when something goes wrong
in this case I don't see any exceptional case